### DISCUSSION TITLE
How to get a pandas dataframe from async result set?

### DISCUSSION THREAD

**Discussion by ghuname** [sqlalchemy/sqlalchemy#7634] [Usage Questions] [answered]
If I have a code like this one (when you run it for the first time comment drop_all line, because the first time table doesn't exist):

```
import asyncio

from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker

from sqlalchemy.orm import declarative_base
from sqlalchemy import Column, Integer, String, DateTime, select
import datetime as dt

Base = declarative_base()

db_path = "sqlite+aiosqlite:///database.sqlite"
engine = create_async_engine(db_path, echo=True, future=True)
async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)


class Book(Base):
    __tablename__ = 'book'

    id = Column(Integer, primary_key=True)
    title = Column(String)
    inserted_at = Column(DateTime, default=dt.datetime.utcnow)

async def main():
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    async with async_session() as session:
        async with session.begin():
            session.add_all(
                [Book(title='A'), Book(title='B')]
            )

    async with async_session() as session:
        stmt = select(Book)
        res = await session.execute(stmt)
        # how to create pandas DataFrame out of res?
        print(res.scalars().all())

loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```

How to get a pandas dataframe out of res (or res.scalars(), or res.scalars().all())?

**Comment by zzzeek** [comment]
pandas would need some method that takes a Result object and produces a dataframe.   I dont know much about pandas and what APIs they offer.  I'm sure there are ways to just make your own dataframe and populate it with the data from a result, though.    as far as async, the normal Result returned by async is buffered so acts the same as the non-async result set (it's the same thing).

**Reply by ghuname** [reply]
I should somehow loop through res.scalars().all() and create list of dictionaries, in dictionary each field is key and each value is field value.
Should I do some object introspection or add a to_dict method to Book class?

```
class Book(Base):
    __tablename__ = 'book'

    id = Column(Integer, primary_key=True)
    title = Column(String)
    inserted_at = Column(DateTime, default=dt.datetime.utcnow)

    def to_dict(self):
        return {
            'id': self.id,
            'title': self.title,
            'inserted_at': self.inserted_at
        }
```

After this addition I can do the following:

```
        res = (await session.execute(stmt)).scalars().all()
        df = pd.DataFrame.from_records([row.to_dict() for row in res])
```

Still, here I have list comprehension which is kind of advanced for loop.
I believe that solution @CaselIT suggested is much better, but I am not sure how to get the sync connection out of async session maker.

**Comment by CaselIT** [comment]
I believe you should use `pandas.read_sql_query` passing the select object and a connection.

Since pandas does not support async, you would need to use `run_sync` to obtain a connection object that has a sync interface and can be used by pandas

**Reply by ghuname** [reply]
> I believe you should use `pandas.read_sql_query` passing the select object and a connection.
> 
> Since pandas does not support async, you would need to use `run_sync` to obtain a connection object that has a sync interface and can be used by pandas

Can you please provide an example of getting the sync connection?

**Reply by CaselIT** [reply]
You can refer to the asyncio documentation https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#running-synchronous-methods-and-functions-under-asyncio

**Reply by ghuname** [reply]
I have read that several times but my knowledge is still shaky. 
For example if I got `session = async_session()` from my original example, may I somehow from that session get async engine like I have executed:

`engine = create_async_engine(db_path, echo=True)`

Or I need two approaches:
1. one with `async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)` for async usage and
2. the other `engine = create_async_engine(db_path, echo=True)`, than 
```
async with engine.begin() as conn:
        await conn.run_sync(something)
```
for non async usages?

It would be quite useful for me if you can provide an example about run_sync and pandas read_sql.

**Reply by CaselIT** [reply, marked as answer]
You can get a connection from a session, with `session.connection()`.

Something like below should work. You can also use a connection directly if you are not using the orm

```py
def pandas_query(session):
  conn = session.connection()
  query = select(a_table)

  return pandas.read_sql_query(query, conn)

async def a_main():
  async with async_session() as session:
    df = await session.run_sync(pandas_query)

```

**Reply by ghuname** [reply]
@CaselIT Yes that is it. Thank you very very much @CaselIT for this solution.
Until pandas invent asyncio, I believe that this is the best solution to the problem.
I have to think about everything now and test it more thoroughly, than I will report back my findings.
Thanks again.

**Reply by einarjohnson** [reply]
@CaselIT , I am running into a very smilar problem. since the `session.connect()` is returning a coroutine object does the `pandas_query` function not need to call await on it? a bit confused on how to puzzle this together.

**Reply by CaselIT** [reply]
The session object passed to the `pandas_query` function above is a sync version, so connection doesn't return a coroutine but a connection.

It's all explained more in the linked docs https://github.com/sqlalchemy/sqlalchemy/discussions/7634#discussioncomment-2073202

**Reply by einarjohnson** [reply]
@CaselIT , of course. I was doing things in a very incorrect way in my code. Thanks for pointing me in the right direction.
